{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Using *fusion-tools* to better understand and annotate spatial --omics data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This cell only needs to be executed once, just making sure that the package is correctly installed \n",
    "# and is accessible by your Python kernel\n",
    "#%pip install --upgrade setuptools\n",
    "#%pip install fusion-tools[interactive]==3.5.71\n",
    "#%pip install large-image[common]==1.29.11"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Make sure you restart the kernel before proceeding!**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Check the output of the following to make sure that the correct package versions were installed:\n",
    "- fusion-tools==3.5.40\n",
    "- large-image==1.29.11"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Package                        VersionNote: you may need to restart the kernel to use updated packages.\n",
      "\n",
      "------------------------------ ------------\n",
      "affine                         2.4.0\n",
      "aiobotocore                    2.5.4\n",
      "aiohappyeyeballs               2.4.0\n",
      "aiohttp                        3.9.3\n",
      "aioitertools                   0.12.0\n",
      "aiosignal                      1.3.1\n",
      "alabaster                      0.7.16\n",
      "anndata                        0.10.6\n",
      "annotated-types                0.7.0\n",
      "ansi2html                      1.8.0\n",
      "anyio                          4.4.0\n",
      "app-model                      0.2.6\n",
      "appdirs                        1.4.4\n",
      "argcomplete                    3.5.3\n",
      "array_api_compat               1.8\n",
      "asciitree                      0.3.3\n",
      "asttokens                      2.4.1\n",
      "attrs                          24.2.0\n",
      "babel                          2.16.0\n",
      "backcall                       0.2.0\n",
      "backports.tarfile              1.2.0\n",
      "beautifulsoup4                 4.12.3\n",
      "bleach                         6.0.0\n",
      "blinker                        1.8.2\n",
      "botocore                       1.31.17\n",
      "build                          1.2.1\n",
      "cachelib                       0.9.0\n",
      "cachetools                     5.5.0\n",
      "cachey                         0.2.1\n",
      "cattrs                         23.2.3\n",
      "certifi                        2024.7.4\n",
      "charset-normalizer             3.3.2\n",
      "click                          8.1.7\n",
      "click-plugins                  1.1.1\n",
      "cligj                          0.7.2\n",
      "cloudpickle                    3.0.0\n",
      "colorama                       0.4.6\n",
      "colorcet                       3.1.0\n",
      "comm                           0.2.2\n",
      "contourpy                      1.2.1\n",
      "ctk-cli                        1.5\n",
      "cycler                         0.12.1\n",
      "dash                           2.18.2\n",
      "dash-bootstrap-components      1.6.0\n",
      "dash-breakpoints               0.1.0\n",
      "dash-core-components           2.0.0\n",
      "dash-cytoscape                 0.3.0\n",
      "dash-draggable                 0.1.2\n",
      "dash-draggable-no-console-logs 0.0.4\n",
      "dash-extensions                1.0.19\n",
      "dash-html-components           2.0.0\n",
      "dash-leaflet                   1.0.15\n",
      "dash_mantine_components        0.14.4\n",
      "dash-svg                       0.0.12\n",
      "dash-table                     5.0.0\n",
      "dash-treeview-antd             0.0.1\n",
      "dash_uploader                  0.7.0a1\n",
      "dask                           2023.9.0\n",
      "dask-image                     2023.8.1\n",
      "dataclass-wizard               0.30.1\n",
      "datashader                     0.16.0\n",
      "debugpy                        1.8.1\n",
      "decorator                      5.1.1\n",
      "defusedxml                     0.7.1\n",
      "dicomweb-client                0.59.1\n",
      "dill                           0.3.7\n",
      "diskcache                      5.6.3\n",
      "distributed                    2023.9.0\n",
      "docopt                         0.6.2\n",
      "docrep                         0.3.2\n",
      "docstring_parser               0.16\n",
      "docutils                       0.20\n",
      "dynamic                        0.1.1\n",
      "EditorConfig                   0.12.4\n",
      "entrypoints                    0.4\n",
      "esbonio                        0.16.4\n",
      "esda                           2.7.0\n",
      "et-xmlfile                     1.1.0\n",
      "executing                      2.0.1\n",
      "fastapi                        0.115.0\n",
      "fasteners                      0.19\n",
      "fastjsonschema                 2.19.1\n",
      "fcsparser                      0.2.8\n",
      "filelock                       3.12.2\n",
      "fiona                          1.9.6\n",
      "Flask                          2.2.5\n",
      "Flask-Caching                  2.3.0\n",
      "fonttools                      4.53.1\n",
      "freetype-py                    2.4.0\n",
      "frozenlist                     1.4.1\n",
      "fsspec                         2024.6.1\n",
      "fusion-tools                   3.5.71\n",
      "geobuf                         1.1.1\n",
      "geojson                        3.1.0\n",
      "geopandas                      1.0.1\n",
      "girder-client                  3.2.8\n",
      "girder-job-sequence            0.2.7\n",
      "h11                            0.14.0\n",
      "h5py                           3.11.0\n",
      "HeapDict                       1.0.1\n",
      "hsluv                          5.0.4\n",
      "idna                           3.7\n",
      "igraph                         0.11.4\n",
      "imagecodecs                    2024.12.30\n",
      "imagecodecs-numcodecs          2024.12.30\n",
      "imageio                        2.35.1\n",
      "imagesize                      1.4.1\n",
      "importlib_metadata             8.4.0\n",
      "in-n-out                       0.2.0\n",
      "inflect                        7.0.0\n",
      "ipykernel                      6.29.3\n",
      "ipython                        8.12.3\n",
      "itsdangerous                   2.2.0\n",
      "jaraco.classes                 3.4.0\n",
      "jaraco.context                 6.0.1\n",
      "jaraco.functools               4.0.2\n",
      "jedi                           0.19.1\n",
      "Jinja2                         3.1.4\n",
      "jmespath                       1.0.1\n",
      "joblib                         1.4.2\n",
      "jsbeautifier                   1.15.1\n",
      "jsonschema                     4.21.1\n",
      "jsonschema-specifications      2023.12.1\n",
      "jupyter_client                 8.6.1\n",
      "jupyter_core                   5.7.2\n",
      "jupyterlab_pygments            0.3.0\n",
      "kaleido                        0.1.0.post1\n",
      "keyring                        24.2.0\n",
      "kiwisolver                     1.4.5\n",
      "lamin_utils                    0.13.1\n",
      "large-image                    1.29.11\n",
      "large-image-source-deepzoom    1.29.11\n",
      "large-image-source-dicom       1.29.11\n",
      "large-image-source-multi       1.29.11\n",
      "large-image-source-nd2         1.29.11\n",
      "large-image-source-pil         1.29.11\n",
      "large-image-source-rasterio    1.29.11\n",
      "large-image-source-test        1.29.11\n",
      "large-image-source-tifffile    1.29.11\n",
      "large-image-source-zarr        1.29.11\n",
      "lazy_loader                    0.4\n",
      "leidenalg                      0.10.2\n",
      "libpysal                       4.12.1\n",
      "llvmlite                       0.43.0\n",
      "locket                         1.0.0\n",
      "loguru                         0.7.2\n",
      "lsprotocol                     2023.0.1\n",
      "lxml                           5.3.1\n",
      "magicgui                       0.8.2\n",
      "markdown-it-py                 3.0.0\n",
      "MarkupSafe                     2.1.5\n",
      "marshmallow                    3.22.0\n",
      "matplotlib                     3.7.2\n",
      "matplotlib-inline              0.1.6\n",
      "matplotlib-scalebar            0.8.1\n",
      "mdurl                          0.1.2\n",
      "mistune                        3.0.2\n",
      "more-itertools                 10.6.0\n",
      "morecantile                    5.3.0\n",
      "mpmath                         1.3.0\n",
      "msgpack                        1.0.8\n",
      "multidict                      6.0.5\n",
      "multipledispatch               1.0.0\n",
      "multiprocess                   0.70.15\n",
      "multiscale_spatial_image       0.11.2\n",
      "napari                         0.4.19.post1\n",
      "napari-console                 0.0.9\n",
      "napari-matplotlib              2.0.1\n",
      "napari-plugin-engine           0.2.0\n",
      "napari-spatialdata             0.4.0\n",
      "napari-svg                     0.1.10\n",
      "natsort                        8.4.0\n",
      "nbclient                       0.10.0\n",
      "nbconvert                      7.16.4\n",
      "nbformat                       5.10.4\n",
      "nd2                            0.10.0\n",
      "nest-asyncio                   1.6.0\n",
      "networkx                       3.3\n",
      "nh3                            0.2.18\n",
      "npe2                           0.7.5\n",
      "numba                          0.60.0\n",
      "numcodecs                      0.13.0\n",
      "numpy                          1.26.4\n",
      "numpydoc                       1.6.0\n",
      "ome-types                      0.5.1\n",
      "ome-zarr                       0.8.3\n",
      "omnipath                       1.0.8\n",
      "opencv-python                  4.8.0.76\n",
      "openpyxl                       3.1.2\n",
      "packaging                      24.1\n",
      "palettable                     3.3.3\n",
      "pandas                         2.2.2\n",
      "pandocfilters                  1.5.1\n",
      "param                          2.1.1\n",
      "parso                          0.8.3\n",
      "partd                          1.4.2\n",
      "patsy                          0.5.6\n",
      "pickleshare                    0.7.5\n",
      "pillow                         10.4.0\n",
      "pillow_heif                    0.18.0\n",
      "pillow-jpls                    1.3.2\n",
      "pillow-jxl-plugin              1.2.6\n",
      "PIMS                           0.6.1\n",
      "Pint                           0.23\n",
      "pip                            25.0.1\n",
      "pipreqs                        0.5.0\n",
      "pipx                           1.7.1\n",
      "pkginfo                        1.10.0\n",
      "platformdirs                   4.2.2\n",
      "plotly                         5.23.0\n",
      "pooch                          1.8.1\n",
      "prompt-toolkit                 3.0.43\n",
      "protobuf                       3.20.0\n",
      "psutil                         6.0.0\n",
      "psygnal                        0.10.2\n",
      "pure-eval                      0.2.2\n",
      "pyarrow                        17.0.0\n",
      "pyconify                       0.1.6\n",
      "pyct                           0.5.0\n",
      "pydantic                       2.10.5\n",
      "pydantic-compat                0.1.2\n",
      "pydantic_core                  2.27.2\n",
      "pydicom                        2.4.4\n",
      "pygeos                         0.14\n",
      "pygls                          1.3.0\n",
      "Pygments                       2.18.0\n",
      "pynndescent                    0.5.11\n",
      "pyogrio                        0.9.0\n",
      "pyopencl                       2024.2.7\n",
      "PyOpenGL                       3.1.7\n",
      "pyparsing                      3.0.9\n",
      "pyproj                         3.6.1\n",
      "pyproject_hooks                1.1.0\n",
      "pyspellchecker                 0.8.1\n",
      "python-dateutil                2.9.0.post0\n",
      "python-multipart               0.0.20\n",
      "pytools                        2024.1.11\n",
      "pytz                           2024.1\n",
      "PyWavelets                     1.4.1\n",
      "pywin32                        306\n",
      "pywin32-ctypes                 0.2.3\n",
      "PyYAML                         6.0.2\n",
      "pyzmq                          25.1.2\n",
      "qtconsole                      5.5.1\n",
      "QtPy                           2.4.1\n",
      "rasterio                       1.3.9\n",
      "rawpy                          0.21.0\n",
      "readfcs                        1.1.7\n",
      "readme-renderer                41.0\n",
      "redis                          5.0.8\n",
      "referencing                    0.35.1\n",
      "requests                       2.32.3\n",
      "requests-toolbelt              1.0.0\n",
      "resource-backed-dask-array     0.1.0\n",
      "retrying                       1.3.4\n",
      "rfc3986                        2.0.0\n",
      "rich                           13.7.1\n",
      "rio-cogeo                      5.2.0\n",
      "rpds-py                        0.20.0\n",
      "ruff                           0.4.10\n",
      "s3fs                           2023.6.0\n",
      "scanpy                         1.9.8\n",
      "scikit-image                   0.24.0\n",
      "scikit-learn                   1.6.1\n",
      "scipy                          1.10.1\n",
      "seaborn                        0.13.2\n",
      "segment-anything               1.0\n",
      "session-info                   1.0.0\n",
      "setuptools                     68.1.2\n",
      "shapely                        2.0.6\n",
      "simplejpeg                     1.7.4\n",
      "six                            1.16.0\n",
      "sleepyjson                     0.0.3\n",
      "slicerator                     1.1.0\n",
      "sniffio                        1.3.1\n",
      "snowballstemmer                2.2.0\n",
      "snuggs                         1.4.7\n",
      "sortedcontainers               2.4.0\n",
      "soupsieve                      2.5\n",
      "spatial_image                  0.3.0\n",
      "spatialdata                    0.1.0\n",
      "spatialdata-io                 0.1.1\n",
      "spatialdata-plot               0.2.0\n",
      "Sphinx                         7.4.7\n",
      "sphinxcontrib-applehelp        2.0.0\n",
      "sphinxcontrib-devhelp          2.0.0\n",
      "sphinxcontrib-htmlhelp         2.1.0\n",
      "sphinxcontrib-jsmath           1.0.1\n",
      "sphinxcontrib-qthelp           2.0.0\n",
      "sphinxcontrib-serializinghtml  2.0.0\n",
      "squidpy                        1.4.1\n",
      "stack-data                     0.6.3\n",
      "starlette                      0.38.2\n",
      "statsmodels                    0.14.2\n",
      "stdlib-list                    0.10.0\n",
      "superqt                        0.6.2\n",
      "sympy                          1.12\n",
      "tabulate                       0.9.0\n",
      "tblib                          3.0.0\n",
      "tenacity                       9.0.0\n",
      "texttable                      1.7.0\n",
      "threadpoolctl                  3.5.0\n",
      "tifffile                       2024.8.10\n",
      "tiffslide                      2.2.0\n",
      "tinycss2                       1.2.1\n",
      "tomli_w                        1.0.0\n",
      "toolz                          0.12.1\n",
      "torch                          2.0.1\n",
      "torchvision                    0.15.2\n",
      "tornado                        6.4.1\n",
      "tqdm                           4.66.5\n",
      "traitlets                      5.14.2\n",
      "transforms3d                   0.4.2\n",
      "twine                          4.0.2\n",
      "typer                          0.10.0\n",
      "typing_extensions              4.12.2\n",
      "tzdata                         2024.1\n",
      "umap-learn                     0.5.7\n",
      "universal_pathlib              0.2.2\n",
      "upload-component               0.0.5\n",
      "urllib3                        1.26.20\n",
      "userpath                       1.9.2\n",
      "uuid                           1.30\n",
      "uvicorn                        0.30.6\n",
      "validators                     0.23.2\n",
      "vispy                          0.14.2\n",
      "waitress                       2.1.2\n",
      "wcwidth                        0.2.13\n",
      "webencodings                   0.5.1\n",
      "Werkzeug                       2.2.3\n",
      "wheel                          0.41.2\n",
      "win32-setctime                 1.1.0\n",
      "wrapt                          1.16.0\n",
      "wsi-annotations-kit            1.4.10\n",
      "wsidicom                       0.20.4\n",
      "xarray                         2024.7.0\n",
      "xarray-dataclasses             1.7.0\n",
      "xarray-datatree                0.0.14\n",
      "xarray-schema                  0.0.3\n",
      "xarray-spatial                 0.3.7\n",
      "xsdata                         24.3.1\n",
      "yarg                           0.1.9\n",
      "yarl                           1.9.7\n",
      "zarr                           2.18.2\n",
      "zict                           3.0.0\n",
      "zipp                           3.20.0\n"
     ]
    }
   ],
   "source": [
    "%pip list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Importing packages as needed\n",
    "\n",
    "In this case, the *Visualization* object and interactive components."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\samuelborder\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\tqdm\\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "import json\n",
    "\n",
    "from fusion_tools.visualization import Visualization\n",
    "from fusion_tools.components import HybridSlideMap, OverlayOptions, PropertyPlotter, BulkLabels, HRAViewer\n",
    "\n",
    "from fusion_tools.utils.omics import get_gene_info\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Pointing to your downloaded files\n",
    "\n",
    "Edit the following lines to indicate where the data you downloaded from the HuBMAP Portal is located on your computer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# In the bottom two lines, os.getcwd() gets the folder that this script is stored in and then that is combined with the names below\n",
    "# to point to the image and annotations\n",
    "downloaded_image = os.path.join(os.getcwd(),'Portal_Data','visium_histology_hires_pyramid.ome.tif')\n",
    "downloaded_analysis = os.path.join(os.getcwd(),'Portal_Data','secondary_analysis.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### If the following lines run successfully then your data paths are valid, if not then check the paths to make sure you're pointing at the right location."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "c:\\Users\\samuelborder\\Desktop\\HIVE_Stuff\\FUSION\\Canvas Course Folder\\Portal_Data\\visium_histology_hires_pyramid.ome.tif\n",
      "True\n",
      "TifffileFileTileSource ('c:\\\\Users\\\\samuelborder\\\\Desktop\\\\HIVE_Stuff\\\\FUSION\\\\Canvas Course Folder\\\\Portal_Data\\\\visium_histology_hires_pyramid.ome.tif', 'JPEG', 95, 0, 'raw', False, '__STYLESTART__', None, '__STYLEEND__')\n"
     ]
    }
   ],
   "source": [
    "print(downloaded_image)\n",
    "\n",
    "print(os.path.exists(downloaded_image))\n",
    "import large_image\n",
    "\n",
    "test = large_image.open(downloaded_image)\n",
    "print(test)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extracting information to add to annotations\n",
    "\n",
    "The following cells just define some functions for getting the genes with the greatest \"dispersion\" value\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_high_dispersion(analysis_data, top_k = 25):\n",
    "\n",
    "    # Finding the \"var\" associated with normalized dispersion and sorting values from high to low\n",
    "    highest_norm_dispersions = analysis_data.var['dispersions_norm'].sort_values(ascending=False)\n",
    "    # Listing just the names and not the values\n",
    "    top_k_names = list(highest_norm_dispersions.index)[:top_k]\n",
    "\n",
    "    return top_k_names\n",
    "\n",
    "def translate_ensg_names(names_list):\n",
    "    # Storing the information in a dictionary\n",
    "    n_info_dict = {}\n",
    "    # Iterating through the provided list of names (tqdm provides a nice progress bar)\n",
    "    for n in tqdm(names_list):\n",
    "        # Using the mygene.info API to extract information on this particular gene ID\n",
    "        n_info = get_gene_info(n)\n",
    "        # Adding information to the dictionary\n",
    "        n_info_dict[n] = n_info\n",
    "\n",
    "    return n_info_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Loading the *secondary_analysis.h5ad* file and pulling out relevant info per-spot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_data = ad.read_h5ad(downloaded_analysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 25/25 [00:17<00:00,  1.44it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"ENSG00000211899.10\": {\n",
      "        \"HGNC\": \"5541\",\n",
      "        \"_id\": \"3507\",\n",
      "        \"_version\": 1,\n",
      "        \"alias\": [\n",
      "            \"AGM1\",\n",
      "            \"MU\",\n",
      "            \"VH\"\n",
      "        ],\n",
      "        \"summary\": \"Immunoglobulins (Ig) are the antigen recognition molecules of B cells. An Ig molecule is made up of 2 identical heavy chains and 2 identical light chains (see MIM 147200) joined by disulfide bonds so that each heavy chain is linked to a light chain and the 2 heavy chains are linked together. Each Ig heavy chain has an N-terminal variable (V) region containing the antigen-binding site and a C-terminal constant (C) region, encoded by an individual C region gene, that determines the isotype of the antibody and provides effector or signaling functions. The heavy chain V region is encoded by 1 each of 3 types of genes: V genes (see MIM 147070), joining (J) genes (see MIM 147010), and diversity (D) genes (see MIM 146910). The C region genes are clustered downstream of the V region genes within the heavy chain locus on chromosome 14. The IGHM gene encodes the C region of the mu heavy chain, which defines the IgM isotype. Naive B cells express the transmembrane forms of IgM and IgD (see IGHD; MIM 1471770) on their surface. During an antibody response, activated B cells can switch to the expression of individual downstream heavy chain C region genes by a process of somatic recombination known as isotype switching. In addition, secreted Ig forms that act as antibodies can be produced by alternative RNA processing of the heavy chain C region sequences. Although the membrane forms of all Ig isotypes are monomeric, secreted IgM forms pentamers, and occasionally hexamers, in plasma (summary by Janeway et al., 2005).[supplied by OMIM, Aug 2010]\"\n",
      "    },\n",
      "    \"ENSG00000211895.5\": {\n",
      "        \"HGNC\": \"5478\",\n",
      "        \"_id\": \"3493\",\n",
      "        \"_version\": 1,\n",
      "        \"alias\": \"IgA1\",\n",
      "        \"summary\": \"Contributes to immunoglobulin receptor binding activity. Involved in antibacterial humoral response; glomerular filtration; and positive regulation of respiratory burst. Located in extracellular space. Part of monomeric IgA immunoglobulin complex and secretory dimeric IgA immunoglobulin complex. [provided by Alliance of Genome Resources, Apr 2022]\"\n",
      "    },\n",
      "    \"ENSG00000211893.4\": {\n",
      "        \"HGNC\": \"5526\",\n",
      "        \"_id\": \"3501\",\n",
      "        \"_version\": 1,\n",
      "        \"summary\": \"Predicted to enable antigen binding activity and immunoglobulin receptor binding activity. Predicted to be involved in several processes, including activation of immune response; defense response to other organism; and phagocytosis. Located in extracellular space. [provided by Alliance of Genome Resources, Apr 2022]\"\n",
      "    },\n",
      "    \"ENSG00000137077.8\": {\n",
      "        \"HGNC\": \"10620\",\n",
      "        \"_id\": \"6366\",\n",
      "        \"_version\": 1,\n",
      "        \"alias\": [\n",
      "            \"6Ckine\",\n",
      "            \"CKb9\",\n",
      "            \"ECL\",\n",
      "            \"SCYA21\",\n",
      "            \"SLC\",\n",
      "            \"TCA4\"\n",
      "        ],\n",
      "        \"summary\": \"This antimicrobial gene is one of several CC cytokine genes clustered on the p-arm of chromosome 9. Cytokines are a family of secreted proteins involved in immunoregulatory and inflammatory processes. The CC cytokines are proteins characterized by two adjacent cysteines. Similar to other chemokines the protein encoded by this gene inhibits hemopoiesis and stimulates chemotaxis. This protein is chemotactic in vitro for thymocytes and activated T cells, but not for B cells, macrophages, or neutrophils. The cytokine encoded by this gene may also play a role in mediating homing of lymphocytes to secondary lymphoid organs. It is a high affinity functional ligand for chemokine receptor 7 that is expressed on T and B lymphocytes and a known receptor for another member of the cytokine family (small inducible cytokine A19). [provided by RefSeq, Sep 2014].\"\n",
      "    },\n",
      "    \"ENSG00000132465.12\": {\n",
      "        \"HGNC\": \"5713\",\n",
      "        \"_id\": \"3512\",\n",
      "        \"_version\": 2,\n",
      "        \"alias\": [\n",
      "            \"IGCJ\",\n",
      "            \"IGJ\",\n",
      "            \"JCH\"\n",
      "        ],\n",
      "        \"summary\": \"Enables IgA binding activity and protein homodimerization activity. Contributes to several functions, including immunoglobulin receptor binding activity; peptidoglycan binding activity; and phosphatidylcholine binding activity. Involved in several processes, including defense response to other organism; glomerular filtration; and positive regulation of respiratory burst. Located in extracellular space. Part of monomeric IgA immunoglobulin complex; pentameric IgM immunoglobulin complex; and secretory dimeric IgA immunoglobulin complex. [provided by Alliance of Genome Resources, Apr 2022]\"\n",
      "    },\n",
      "    \"ENSG00000244734.4\": {\n",
      "        \"HGNC\": \"4827\",\n",
      "        \"_id\": \"3043\",\n",
      "        \"_version\": 2,\n",
      "        \"alias\": [\n",
      "            \"CD113t-C\",\n",
      "            \"ECYT6\",\n",
      "            \"beta-globin\"\n",
      "        ],\n",
      "        \"summary\": \"The alpha (HBA) and beta (HBB) loci determine the structure of the 2 types of polypeptide chains in adult hemoglobin, Hb A. The normal adult hemoglobin tetramer consists of two alpha chains and two beta chains. Mutant beta globin causes sickle cell anemia. Absence of beta chain causes beta-zero-thalassemia. Reduced amounts of detectable beta globin causes beta-plus-thalassemia. The order of the genes in the beta-globin cluster is 5'-epsilon -- gamma-G -- gamma-A -- delta -- beta--3'. [provided by RefSeq, Jul 2008].\"\n",
      "    },\n",
      "    \"ENSG00000146674.15\": {\n",
      "        \"HGNC\": \"5472\",\n",
      "        \"_id\": \"3486\",\n",
      "        \"_version\": 1,\n",
      "        \"alias\": [\n",
      "            \"BP-53\",\n",
      "            \"IBP-3\",\n",
      "            \"IBP3\",\n",
      "            \"IGFBP-3\"\n",
      "        ],\n",
      "        \"summary\": \"This gene is a member of the insulin-like growth factor binding protein (IGFBP) family and encodes a protein with an IGFBP domain and a thyroglobulin type-I domain. The protein forms a ternary complex with insulin-like growth factor acid-labile subunit (IGFALS) and either insulin-like growth factor (IGF) I or II. In this form, it circulates in the plasma, prolonging the half-life of IGFs and altering their interaction with cell surface receptors. Alternate transcriptional splice variants, encoding different isoforms, have been characterized. [provided by RefSeq, Jul 2008].\"\n",
      "    },\n",
      "    \"ENSG00000206172.8\": {\n",
      "        \"HGNC\": \"4823\",\n",
      "        \"_id\": \"3039\",\n",
      "        \"_version\": 2,\n",
      "        \"alias\": [\n",
      "            \"ECYT7\",\n",
      "            \"HBA-T3\",\n",
      "            \"HBH\",\n",
      "            \"METHBA\"\n",
      "        ],\n",
      "        \"summary\": \"The human alpha globin gene cluster located on chromosome 16 spans about 30 kb and includes seven loci: 5'- zeta - pseudozeta - mu - pseudoalpha-1 - alpha-2 - alpha-1 - theta - 3'. The alpha-2 (HBA2) and alpha-1 (HBA1) coding sequences are identical. These genes differ slightly over the 5' untranslated regions and the introns, but they differ significantly over the 3' untranslated regions. Two alpha chains plus two beta chains constitute HbA, which in normal adult life comprises about 97% of the total hemoglobin; alpha chains combine with delta chains to constitute HbA-2, which with HbF (fetal hemoglobin) makes up the remaining 3% of adult hemoglobin. Alpha thalassemias result from deletions of each of the alpha genes as well as deletions of both HBA2 and HBA1; some nondeletion alpha thalassemias have also been reported. [provided by RefSeq, Jul 2008].\"\n",
      "    },\n",
      "    \"ENSG00000188536.13\": {\n",
      "        \"HGNC\": \"4824\",\n",
      "        \"_id\": \"3040\",\n",
      "        \"_version\": 2,\n",
      "        \"alias\": [\n",
      "            \"ECYT7\",\n",
      "            \"HBA-T2\",\n",
      "            \"HBH\"\n",
      "        ],\n",
      "        \"summary\": \"The human alpha globin gene cluster located on chromosome 16 spans about 30 kb and includes seven loci: 5'- zeta - pseudozeta - mu - pseudoalpha-1 - alpha-2 - alpha-1 - theta - 3'. The alpha-2 (HBA2) and alpha-1 (HBA1) coding sequences are identical. These genes differ slightly over the 5' untranslated regions and the introns, but they differ significantly over the 3' untranslated regions. Two alpha chains plus two beta chains constitute HbA, which in normal adult life comprises about 97% of the total hemoglobin; alpha chains combine with delta chains to constitute HbA-2, which with HbF (fetal hemoglobin) makes up the remaining 3% of adult hemoglobin. Alpha thalassemias result from deletions of each of the alpha genes as well as deletions of both HBA2 and HBA1; some nondeletion alpha thalassemias have also been reported. [provided by RefSeq, Jul 2008].\"\n",
      "    },\n",
      "    \"ENSG00000188257.11\": {\n",
      "        \"HGNC\": \"9031\",\n",
      "        \"_id\": \"5320\",\n",
      "        \"_version\": 1,\n",
      "        \"alias\": [\n",
      "            \"MOM1\",\n",
      "            \"PLA2\",\n",
      "            \"PLA2B\",\n",
      "            \"PLA2L\",\n",
      "            \"PLA2S\",\n",
      "            \"PLAS1\",\n",
      "            \"sPLA2\"\n",
      "        ],\n",
      "        \"summary\": \"The protein encoded by this gene is a member of the phospholipase A2 family (PLA2). PLA2s constitute a diverse family of enzymes with respect to sequence, function, localization, and divalent cation requirements. This gene product belongs to group II, which contains secreted form of PLA2, an extracellular enzyme that has a low molecular mass and requires calcium ions for catalysis. It catalyzes the hydrolysis of the sn-2 fatty acid acyl ester bond of phosphoglycerides, releasing free fatty acids and lysophospholipids, and thought to participate in the regulation of the phospholipid metabolism in biomembranes. Several alternatively spliced transcript variants with different 5' UTRs have been found for this gene.[provided by RefSeq, Sep 2009].\"\n",
      "    },\n",
      "    \"ENSG00000211898.7\": {\n",
      "        \"HGNC\": \"5480\",\n",
      "        \"_id\": \"3495\",\n",
      "        \"_version\": 1,\n",
      "        \"summary\": \"Predicted to enable antigen binding activity and immunoglobulin receptor binding activity. Involved in positive regulation of interleukin-1 production. Located in blood microparticle and extracellular exosome. [provided by Alliance of Genome Resources, Apr 2022]\"\n",
      "    },\n",
      "    \"ENSG00000286848.1\": {\n",
      "        \"_id\": \"101927297\",\n",
      "        \"_version\": 1\n",
      "    },\n",
      "    \"ENSG00000211890.4\": {\n",
      "        \"HGNC\": \"5479\",\n",
      "        \"_id\": \"3494\",\n",
      "        \"_version\": 1,\n",
      "        \"alias\": \"A2M\",\n",
      "        \"summary\": \"Contributes to immunoglobulin receptor binding activity. Involved in antibacterial humoral response; glomerular filtration; and positive regulation of respiratory burst. Located in extracellular space. Part of monomeric IgA immunoglobulin complex and secretory dimeric IgA immunoglobulin complex. [provided by Alliance of Genome Resources, Apr 2022]\"\n",
      "    },\n",
      "    \"ENSG00000253755.1\": {\n",
      "        \"_id\": \"ENSG00000253755\",\n",
      "        \"_version\": 1\n",
      "    },\n",
      "    \"ENSG00000163220.11\": {\n",
      "        \"HGNC\": \"10499\",\n",
      "        \"_id\": \"6280\",\n",
      "        \"_version\": 1,\n",
      "        \"alias\": [\n",
      "            \"60B8AG\",\n",
      "            \"CAGB\",\n",
      "            \"CFAG\",\n",
      "            \"CGLB\",\n",
      "            \"L1AG\",\n",
      "            \"LIAG\",\n",
      "            \"MAC387\",\n",
      "            \"MIF\",\n",
      "            \"MRP14\",\n",
      "            \"NIF\",\n",
      "            \"P14\",\n",
      "            \"S100-A9\"\n",
      "        ],\n",
      "        \"summary\": \"The protein encoded by this gene is a member of the S100 family of proteins containing 2 EF-hand calcium-binding motifs. S100 proteins are localized in the cytoplasm and/or nucleus of a wide range of cells, and involved in the regulation of a number of cellular processes such as cell cycle progression and differentiation. S100 genes include at least 13 members which are located as a cluster on chromosome 1q21. This protein may function in the inhibition of casein kinase and altered expression of this protein is associated with the disease cystic fibrosis. This antimicrobial protein exhibits antifungal and antibacterial activity. [provided by RefSeq, Nov 2014].\"\n",
      "    },\n",
      "    \"ENSG00000236552.2\": {\n",
      "        \"HGNC\": \"23736\",\n",
      "        \"_id\": \"728658\",\n",
      "        \"_version\": 1,\n",
      "        \"alias\": [\n",
      "            \"RPL13A_8_1086\",\n",
      "            \"bA196N24.2\"\n",
      "        ]\n",
      "    },\n",
      "    \"ENSG00000072849.11\": {\n",
      "        \"HGNC\": \"17943\",\n",
      "        \"_id\": \"51009\",\n",
      "        \"_version\": 1,\n",
      "        \"alias\": [\n",
      "            \"CGI-101\",\n",
      "            \"DERtrin-2\",\n",
      "            \"F-LAN-1\",\n",
      "            \"F-LANa\",\n",
      "            \"FLANa\",\n",
      "            \"derlin-2\"\n",
      "        ],\n",
      "        \"summary\": \"Proteins that are unfolded or misfolded in the endoplasmic reticulum (ER) must be refolded or degraded to maintain the homeostasis of the ER. DERL2 is involved in the degradation of misfolded glycoproteins in the ER (Oda et al., 2006 [PubMed 16449189]).[supplied by OMIM, Mar 2008].\"\n",
      "    },\n",
      "    \"ENSG00000205364.4\": {\n",
      "        \"HGNC\": \"14296\",\n",
      "        \"_id\": \"4499\",\n",
      "        \"_version\": 1,\n",
      "        \"alias\": [\n",
      "            \"MT-1M\",\n",
      "            \"MT-IM\",\n",
      "            \"MT1\",\n",
      "            \"MT1K\"\n",
      "        ],\n",
      "        \"summary\": \"This gene encodes a member of the metallothionein superfamily, type 1 family. Metallothioneins have a high content of cysteine residues that bind various heavy metals. These genes are transcriptionally regulated by both heavy metals and glucocorticoids. [provided by RefSeq, Oct 2011].\"\n",
      "    },\n",
      "    \"ENSG00000211679.2\": {\n",
      "        \"HGNC\": \"5857\",\n",
      "        \"_id\": \"3539\",\n",
      "        \"_version\": 1,\n",
      "        \"alias\": \"IGLC\",\n",
      "        \"summary\": \"Predicted to enable antigen binding activity and immunoglobulin receptor binding activity. Predicted to be involved in several processes, including activation of immune response; defense response to other organism; and phagocytosis. Located in blood microparticle and extracellular exosome. [provided by Alliance of Genome Resources, Apr 2022]\"\n",
      "    },\n",
      "    \"ENSG00000254709.8\": {\n",
      "        \"HGNC\": \"38476\",\n",
      "        \"_id\": \"100423062\",\n",
      "        \"_version\": 1,\n",
      "        \"alias\": [\n",
      "            \"IGL\",\n",
      "            \"IGLV\",\n",
      "            \"VL-MAR\"\n",
      "        ],\n",
      "        \"summary\": \"This gene encodes one of the immunoglobulin lambda-like polypeptides. It is located within the immunoglobulin lambda locus but it does not require somatic rearrangement for expression. The first exon of this gene is unrelated to immunoglobulin variable genes; the second and third exons are the immunoglobulin lambda joining 1 and the immunoglobulin lambda constant 1 gene segments. Alternative splicing results in multiple transcript variants. [provided by RefSeq, May 2010].\"\n",
      "    },\n",
      "    \"ENSG00000240040.6\": {\n",
      "        \"_id\": \"ENSG00000240040\",\n",
      "        \"_version\": 1\n",
      "    },\n",
      "    \"ENSG00000276409.5\": {\n",
      "        \"HGNC\": \"10612\",\n",
      "        \"_id\": \"6358\",\n",
      "        \"_version\": 1,\n",
      "        \"alias\": [\n",
      "            \"CC-1\",\n",
      "            \"CC-3\",\n",
      "            \"CKB1\",\n",
      "            \"HCC-1\",\n",
      "            \"HCC-1(1-74)\",\n",
      "            \"HCC-1/HCC-3\",\n",
      "            \"HCC-3\",\n",
      "            \"MCIF\",\n",
      "            \"NCC-2\",\n",
      "            \"NCC2\",\n",
      "            \"SCYA14\",\n",
      "            \"SCYL2\",\n",
      "            \"SY14\"\n",
      "        ],\n",
      "        \"summary\": \"This gene, chemokine (C-C motif) ligand 14, is one of several CC cytokine genes clustered on 17q11.2. The CC cytokines are secreted proteins characterized by two adjacent cysteines. The cytokine encoded by this gene induces changes in intracellular calcium concentration and enzyme release in monocytes. Multiple transcript variants encoding different isoforms have been found for this gene. Read-through transcripts are also expressed that include exons from the upstream cytokine gene, chemokine (C-C motif) ligand 15, and are represented as GeneID: 348249. [provided by RefSeq, Dec 2009].\"\n",
      "    },\n",
      "    \"ENSG00000211459.2\": {\n",
      "        \"HGNC\": \"7470\",\n",
      "        \"_id\": \"4549\",\n",
      "        \"_version\": 1,\n",
      "        \"alias\": \"MTRNR1\",\n",
      "        \"summary\": \"Enables DNA binding activity and DNA-binding transcription factor binding activity. Involved in several processes, including osteoblast proliferation; regulation of carbohydrate utilization; and regulation of phosphate metabolic process. Located in extracellular space; mitochondrion; and nucleus. [provided by Alliance of Genome Resources, Apr 2022]\"\n",
      "    },\n",
      "    \"ENSG00000160613.14\": {\n",
      "        \"HGNC\": \"8748\",\n",
      "        \"_id\": \"9159\",\n",
      "        \"_version\": 1,\n",
      "        \"alias\": [\n",
      "            \"LPC\",\n",
      "            \"PC7\",\n",
      "            \"PC8\",\n",
      "            \"SPC7\"\n",
      "        ],\n",
      "        \"summary\": \"This gene encodes a member of the subtilisin-like proprotein convertase family, which includes proteases that process protein and peptide precursors trafficking through regulated or constitutive branches of the secretory pathway. It encodes a type 1 membrane bound protease that is expressed in many tissues, including neuroendocrine, liver, gut, and brain. The encoded protein undergoes an initial autocatalytic processing event in the ER and then sorts to the trans-Golgi network through endosomes where a second autocatalytic event takes place and the catalytic activity is acquired. This gene encodes one of the seven basic amino acid-specific members which cleave their substrates at single or paired basic residues. It can process proalbumin and is thought to be responsible for the activation of HIV envelope glycoproteins gp160 and gp140. This gene has been implicated in the transcriptional regulation of housekeeping genes and plays a role in the regulation of iron metabolism. A t(11;14)(q23;q32) chromosome translocation associated with B-cell lymphoma occurs between this gene and its inverted counterpart. [provided by RefSeq, Feb 2014].\"\n",
      "    },\n",
      "    \"ENSG00000211896.7\": {\n",
      "        \"HGNC\": \"5525\",\n",
      "        \"_id\": \"3500\",\n",
      "        \"_version\": 1,\n",
      "        \"summary\": \"Predicted to enable antigen binding activity and immunoglobulin receptor binding activity. Predicted to be involved in several processes, including activation of immune response; defense response to other organism; and phagocytosis. Predicted to act upstream of or within several processes, including immunoglobulin mediated immune response; positive regulation of hypersensitivity; and positive regulation of phagocytosis. Located in extracellular space. [provided by Alliance of Genome Resources, Apr 2022]\"\n",
      "    }\n",
      "}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "top_k_variable = extract_high_dispersion(analysis_data)\n",
    "top_k_info = translate_ensg_names(top_k_variable)\n",
    "\n",
    "# This is a nice way of visualizing the extracted gene information. \n",
    "# Any that don't have an \"alias\" field just aren't well defined compared to the others that have a summary and alias\n",
    "print(json.dumps(top_k_info,indent=4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(json.dumps(top_k_info[list(top_k_info.keys())[0]],indent=4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Now creating the Spot annotations and adding the most variable gene info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loaded annotations with the gene names found in the previous step\n",
    "from fusion_tools.utils.shapes import load_visium\n",
    "downloaded_annotations = load_visium(downloaded_analysis,include_var_names = top_k_variable)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Setting up the visualization \n",
    "In this next section, we'll define a *Visualization* object including a *SlideMap*, *OverlayOptions*, *PropertyPlotter*, *HRAViewer*, and *BulkLabels*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Defining a Visualization object with images, annotations, and components\n",
    "vis_object = Visualization(\n",
    "    local_slides = [downloaded_image],\n",
    "    local_annotations = [downloaded_annotations],\n",
    "    linkage = 'row',\n",
    "    components = [\n",
    "        [\n",
    "            HybridSlideMap(),\n",
    "            [\n",
    "                OverlayOptions(),\n",
    "                PropertyPlotter(),\n",
    "                HRAViewer(),\n",
    "                BulkLabels()\n",
    "            ]\n",
    "        ]\n",
    "    ],\n",
    "    app_options = {\n",
    "        'jupyter': True,\n",
    "        'default_page': 'main'\n",
    "    }\n",
    ")\n",
    "\n",
    "# This line starts the visualization \n",
    "vis_object.start()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
